lib/hsa/mcwamp_hsa.cpp

```diff
@@ -763,11 +767,56 @@ pool_iterator::pool_iterator()
 ///
 namespace Kalmar {
 
+
+
+// Small wrapper around the hsa hardware queue (ie returned from hsa_queue_create(...).
+// This allows us to see which accelerator_view owns the hsa queue, and
+// also tracks the state of the cu mask, profiling, priority of the HW queue.
+struct RocrQueue {
+    RocrQueue(hsa_queue_t *hwQueue, HSAQueue *hccQueue) : 
+        _hwQueue(hwQueue), _hccQueue(hccQueue) 
+    {
+        assignHccQueue(hccQueue);
+    }
+
+    ~RocrQueue() {
+
+#if KALMAR_DEBUG
+        std::cerr << "HSAQueue::dispose(): destroy an HSA command queue: " << _hwQueue << "\n";
+#endif
+
+        hsa_status_t status = hsa_queue_destroy(_hwQueue);
+        STATUS_CHECK(status, __LINE__);
+    };
+
+    void assignHccQueue(HSAQueue *hccQueue);
+
+    hsa_status_t setCuMask(HSAQueue *hccQueue);
+
+
+    hsa_queue_t *_hwQueue; // Pointer to the HSA queue this entry tracks.
+
+    HSAQueue *_hccQueue;  // Pointe to the HCC "HSA" queue which is assigned to use the rocrQueue
+
+    std::vector<uint32_t> cu_arrays;
+    
+    // Track profiling enabled state here. - no need now since all hw queues have profiling enabled.
+    
+    // Priority could be tracked here:
+};
+
+
+
 class HSAQueue final : public KalmarQueue
 {
 private:
-    // HSA commmand queue associated with this HSAQueue instance
-    hsa_queue_t* commandQueue;
+    friend class Kalmar::HSADevice;
+    friend class RocrQueue;
+
+    // ROCR queue associated with this HSAQueue instance. 
+    RocrQueue    *rocrQueue;
```

Comment:
Yes.  
Really I think we should rename HSAQueue to something like HSAAcceleratorView - this is actually the implementation of accelerator_view for the "hsa" backend.  
Then RocrQueue could be HSAQueue, this is thin wrapper around the hsa_queue_t.